lib/lib_sms.php

```diff
@@ -105,9 +112,107 @@ function sms_getShiftChangeContacts(&$contacts, $day, $startTime, $endTime, $bou
         return false;
     }
 
+    // If looking for shift endings and the specified end time is midnight,
+    // then a check for next day continuation of shifts will need to be made
+    // in order to ensure that notifications do not get sent for shifts that
+    // are continuing through midnight.
+    if ((($boundary == "latest") && ($endTime == "24:00")) ||
+            (($boundary == "earliest") && ($startTime == "00:00"))) {
+
+        // See if any shifts that were retrieved end at or start at midnight;
+        // if not, there is no need to do any further checking.
+        $contactShiftIndicesToCheck = array();
+        foreach ($contacts as $index => $contact) {
+            if ($contact[$boundary] == ($boundary == "latest" ? "23:59:00" : "00:00:00")) {
+                $contactShiftIndicesToCheck[] = $index;
+            }
+        }
+        if (empty($contactShiftIndicesToCheck)) {
+            return true;
+        }
+
+        // Query the database for shifts that are starting at midnight the
+        // next day (if this invocation is to find shifts that are ending)
+        // or that ended the previous day (if this invocation is to find
+        // shifts that are starting).
+        if ($boundary == "latest") {
+            $otherBoundary = "earliest";
+            $otherDay = date("D", strtotime($day."+1 day"));
+            $otherTime = "00:00";
+        } else {
+            $otherBoundary = "latest";
+            $otherDay = date("D", strtotime($day."-1 day"));
+            $otherTime = "23:59";
+        }
+        $otherDayWeekend = ($otherDay == 'Sun' || $otherDay == 'Sat');
+        $otherSql = "SELECT DISTINCT contacts.*, earliest, latest, ".
+            "receive_texts, receive_calls, receive_call_answered_alerts ".
+            "FROM call_times LEFT JOIN contacts ".
+            "ON contacts.id = call_times.contact_id ".
+            "WHERE enabled='y' AND ". "(day='all' OR day='{$otherDay}' OR ".
+            ($otherDayWeekend ? "day='weekends'" : "day='weekdays'") . ") AND ".
+            $otherBoundary."='".$otherTime."'";
+        if (!db_db_query($otherSql, $otherContacts, $error)) {
+            return false;
+        }
+
+        // Create a set of the shifts at the other side of the midnight
+        // boundary, with each element being a concatenation of the
+        // contact's name and the three communication type flags for
+        // that shift. This will be used below to match against the
+        // shift change entries that are to be pruned of any that are
+        // not really ending or starting because they are contiguous
+        // with these "other side of midnight" shifts.
+        $otherContactNamesAndCommTypeFlags = array();
+        foreach ($otherContacts as $contact) {
+            $otherContactNamesAndCommTypeFlags[sms_getCallTimeEntryAsKey($contact)] = true;
+        }
+
+
+        echo "Base contacts to be checked:\n";
```

Comment:
Did you intend to leave these "echo" commands in here?  These will be logged with every cron job, right?